src/std_seq.md

```diff
@@ -0,0 +1,148 @@
+
+Here are the function signatures of the sequential functions that `@std/seq` will provide:
+
+```ln
+fn next(seq: Seq): Result<int64>
+fn each(seq: Seq, func: function): void
+fn while(seq: Seq, condFn: function, bodyFn: function): void
+fn doWhile(seq: Seq, bodyFn: function): void
+fn recurse(seq: Seq, recursiveFn: function, arg: any): anythingElse
+```
+
+### `Seq` type
+
+`@std/seq` provides functions for accomplishing many different sequential patterns that depend on a special, built-in `Seq` type with opaque internals to avoid manipulation after construction:
+
+```alan
+type Seq {
+  counter: int64
+  limit: int64
+}
+```
+
+### `next_loop.ln`
+
+```rust,editable
+from @std/app import start, print, exit
+from @std/seq import seq, next
+on start {
+  let s = seq(2)
+  print(s.next())
+  print(s.next())
+  print(s.next())
+  emit exit 0
+}
+```
+
+`next` is the simplest to explain: each time you call it, it returns the current counter value wrapped in a Result and then increments it. If you call past the limit, it returns an [Error Result](./error_handling.md).
+
+### `each_loop.ln`
+
+```rust,editable
+from @std/app import start, print, exit
+from @std/seq import seq, each
+on start {
+  let s = seq(3)
+  s.each(fn (i: int64) = print(i))
+  emit exit 0
+}
+```
+
+`each` is almost as simple as `next`: It simply runs the side-effect function however many times is in the limit of the seq instance. The function provided may or may not take the current iteration counter and must be the following signature:
+
+```alan
+fn func(): void
+fn func(i: int64): void
+```
+
+### `while_loop.ln`
+
+```rust,editable
+from @std/app import start, print, exit
+from @std/seq import seq, while
+on start {
+  let s = seq(100)
+  let sum = 0
+  // TODO: Still need type inference working for one-liner closures
+  // https://github.com/alantech/alan/issues/193
+  s.while(fn (): bool = sum < 10, fn {
+    sum = sum + 1
+  })
+  print(sum)
+  emit exit 0
+}
+```
+
+`while` runs the `bodyFn` *up to* the `limit` number of times, but can abort early if `condFn` returns `false`. The signatures of these two functions must match:
+
+```alan
+fn condFn(): bool
+fn bodyFn(): void
+```
+
+### `doWhile_loop.ln`
+
+```rust,editable
+from @std/app import start, print, exit
+from @std/seq import seq, doWhile
+on start {
+  let s = seq(100)
+  let sum = 0
+  // TODO: Still need type inference working for one-liner closures
+  // https://github.com/alantech/alan/issues/193
+  s.doWhile(fn (): bool {
+    sum = sum + 1
+    return sum < 10
+  })
+  print(sum)
+  emit exit 0
+}
+```
+
+`doWhile` always runs at least once (unless the `seq` has reached its `limit` or it was constructed with an initial `limit` of `0`) and uses the return value of the function to determine if it should continue or not, so its `bodyFn` has the following signature:
+
+```alan
+fn bodyFn(): bool
+```
+
+### `recurse_fib.ln`
```

Comment:
Maybe point out that this is not implemented in the AVM yet but only the js-runtime?